# Assignment 1

![](https://media.giphy.com/media/xT9C25UNTwfZuk85WP/giphy-downsized-large.gif)

Remember the rules of ~Fight~ Code Club:
1. ALWAYS DOCUMENT
2. Cite resources that you use (paste links)
3. Include the names people who you worked with
4. Be neat and organized

## Scrape and Clean data

Based on you proposal, scrape or collect your data:

1. One variable must be either: (40 pts)
    1. scraped from the web OR;
    2. collected from an API AND you must create one new variable that is "new" to the best of your knowledge (combination of other variables representing something new).

2. You must have at least 3 variables, but you may include as many as you want into you final dataset. Likely, you will want to include more to make graphs and regressions. (30pts)

3. You must *be able* to run a regression that makes some sense with this data (the regression doesn't have to be a complete model). Briefly describe one regression you would run with your variables. (**DO NOT** run a regression, yet). (15pts)

4. You must have one combined and cleaned dataset (15 pts)

You must submit one python notebook on how you scraped/gathered data from an api, and how you combined and cleaned you data. I should be able to run your code and reproduce your final data set.  

The other variables that you choose to include do not have to be collected by API or webscraped, but you do have to combine the files and clean the dataset with python.

Thus, you must submit:
- Your finalized data set (only one) (note: you may add more variables in the future).
- Your documented python notebook
- Any associated data files needed to produce the final dataset.

You will be evaluated on:
- Completeness of the data
- Quality of the code 
- The creativity of the new variable/webscraped data you gather

Be sure to upload ALL associated files for your code to run. I will run your code from beginning to end - make it easy for me to replicate your code.

In [2]:
import requests
import numpy as np
import pandas as pd
from bs4 import BeautifulSoup  #this is the spell that helps you read tags"
# from pathlib import Path, PureWindowsPath # For windows only 
import os 
import pandas as pd
import time
from time import sleep
import selenium #Need this step 
from selenium import webdriver #Need this step 

from selenium.webdriver.common.by import By #Allows for selenium to click things 
from selenium.webdriver.chrome.service import Service #https://stackoverflow.com/questions/64717302/deprecationwarning-executable-path-has-been-deprecated-selenium-python
from selenium.webdriver.support import expected_conditions as EC #Allows for more complex code 
from selenium.webdriver.chrome.options import Options #Allows you to change aspects of the browser

#importing everything I need

In [3]:
mlb_url = 'https://www.mlb.com/stats/team/ground-out-fly-out-ratio/2025?expanded=true'
# Use requests to retrieve data from a given URL
mlb_response = requests.get(mlb_url)
# Parse the whole HTML page using BeautifulSoup
mlb_soup = BeautifulSoup(mlb_response.text, 'html.parser')
    
# teams in mlb
teams = [] # empty list
tags = mlb_soup.find_all('td', {'class':'col-group-end-BOW7diD7 number-GoaicxKV align-left-L6MdxTlJ is-table-pinned-lGP8KWTK'})  #get the tags for the teams


for link in tags: # for each link in the list tags
    x = link.get_text()  # get the text within the node in that list item link
    teams.append(x) #append it to the list called teams
    
#webscrapping teams but ended up doing it a different way and not using this list 
    

In [5]:
len(teams)

30

In [6]:
#players name in mlb
players = [] #creates empty list called players
for i in range(1, 32):
    mlb_url1 = f"https://www.mlb.com/stats/2025?page={i}" #loops through all 31 pages on mlb site

# Use requests to retrieve data from a given URL
    mlb_response1 = requests.get(mlb_url1)
# Parse the whole HTML page using BeautifulSoup
    mlb_soup1 = BeautifulSoup(mlb_response1.text, 'html.parser')

 # find all player links
    player_links = mlb_soup1.find_all('a', class_='bui-link') #find all a tags on the page that have class bui-link

    for link in player_links: #loops through each tag found 
        # find first and last name spans inside this <a>
        spans = link.find_all('span', class_='full-G_bAyq40') #finds all elemnets with class specified
        if len(spans) >= 2:  # make sure we have first and last name (2 elements in link
            first_name = spans[0].get_text(strip=True)
            last_name = spans[1].get_text(strip=True)
            #indicated first span is first name and second span is last name
            full_name = f"{first_name} {last_name}" #creates 1 variable with first and last name
            players.append(full_name) #appends names to a list called players


In [30]:
players[0] #check that it worked 

'Cal Raleigh'

In [9]:
# homeruns 


homers = [] #initalizes an empty list called homers
for i in range(1, 32):
    mlb_url1 = f"https://www.mlb.com/stats/2025?page={i}" #loops through all pages on site

# Use requests to retrieve data from a given URL
    mlb_response2 = requests.get(mlb_url1)
# Parse the whole HTML page using BeautifulSoup
    mlb_soup2 = BeautifulSoup(mlb_response2.text, 'html.parser')

    tags1 = mlb_soup2.find_all('td', {'class':'selected-h6IPIIxg number-GoaicxKV align-right-TwjGe_gi is-table-pinned-lGP8KWTK'})  #get the tags for the teams
#tags1 is a list of all td specified elements 

    for link in tags1: # for each link in the list tags
        x1 = link.get_text()  # get the text within the node in that list item link
        homers.append(x1) #append it to the list called homers


In [31]:
homers[0]

'60'

In [15]:
# rbis 
#I needed selenium for this because the website code was in java
chrome_options = Options()
driver = webdriver.Chrome() # Initialize
chrome_options.add_argument("--window-size=1900,1000") #pick a size and stick with it, it can change how code intreacts with the website.
driver = webdriver.Chrome(options = chrome_options) #establish driver
#__________________________________________________________________________________________________#
mlb_urls = ["https://www.mlb.com/stats/2025", "https://www.mlb.com/stats/2025?page=2", "https://www.mlb.com/stats/2025?page=3", 
               "https://www.mlb.com/stats/2025?page=4", "https://www.mlb.com/stats/2025?page=5", "https://www.mlb.com/stats/2025?page=6", 
               "https://www.mlb.com/stats/2025?page=7", "https://www.mlb.com/stats/2025?page=8", "https://www.mlb.com/stats/2025?page=9",
               "https://www.mlb.com/stats/2025?page=10", "https://www.mlb.com/stats/2025?page=11", "https://www.mlb.com/stats/2025?page=12",
               "https://www.mlb.com/stats/2025?page=13", "https://www.mlb.com/stats/2025?page=14", "https://www.mlb.com/stats/2025?page=15", "https://www.mlb.com/stats/2025?page=16",
               "https://www.mlb.com/stats/2025?page=17", "https://www.mlb.com/stats/2025?page=18", "https://www.mlb.com/stats/2025?page=19", "https://www.mlb.com/stats/2025?page=20",
               "https://www.mlb.com/stats/2025?page=21", "https://www.mlb.com/stats/2025?page=22", "https://www.mlb.com/stats/2025?page=23", "https://www.mlb.com/stats/2025?page=24",
               "https://www.mlb.com/stats/2025?page=25", "https://www.mlb.com/stats/2025?page=26", "https://www.mlb.com/stats/2025?page=27", "https://www.mlb.com/stats/2025?page=28",
               "https://www.mlb.com/stats/2025?page=29", "https://www.mlb.com/stats/2025?page=30", "https://www.mlb.com/stats/2025?page=31"]
#list of all the pages that I needed to loop through to get my data

from selenium.webdriver.common.by import By

mlb_rbis = [] #empty list called mpb_rbis

for url in mlb_urls: #loops through each url in list above
    driver.get(url) #Uses Selenium’s Chrome WebDriver to open the page in a browser
    driver.implicitly_wait(7) #Tells Selenium to wait up to 7 seconds for elements to load before throwing an error
    
    rows = driver.find_elements(By.XPATH, '//*[@id="stats-app-root"]/section/section/div[4]/div[1]/div/table/tbody/tr')
    #Finds all table rows t in the stats table on the page. Returns row elements 
    for row in rows: #loops through each row in table (each player)
        try:
            rbi = row.find_element(By.XPATH, './td[9]').text #Looks inside the current row (.) for the 9th <td> element (the rbi column) and extracts text
            mlb_rbis.append(int(rbi)) #Converts the RBI value from a string to an integer and adds the number to mlb_rbi list
        except:
            continue
#Inside a try block to avoid errors if a row doesn’t have an RBI value. 
# except/continue: If there’s an error (e.g., RBI column missing, text not an integer), it skips that row and continues to the next row.

In [17]:
mlb_rbis[0] #check it worked

125

In [19]:
# average 
chrome_options = Options()
driver = webdriver.Chrome() # Initialize
chrome_options.add_argument("--window-size=1900,1000") #pick a size and stick with it, it can change how code intreacts with the website.
driver = webdriver.Chrome(options = chrome_options) #establish driver
#__________________________________________________________________________________________________#
mlb_urls = ["https://www.mlb.com/stats/2025", "https://www.mlb.com/stats/2025?page=2", "https://www.mlb.com/stats/2025?page=3", 
               "https://www.mlb.com/stats/2025?page=4", "https://www.mlb.com/stats/2025?page=5", "https://www.mlb.com/stats/2025?page=6", 
               "https://www.mlb.com/stats/2025?page=7", "https://www.mlb.com/stats/2025?page=8", "https://www.mlb.com/stats/2025?page=9",
               "https://www.mlb.com/stats/2025?page=10", "https://www.mlb.com/stats/2025?page=11", "https://www.mlb.com/stats/2025?page=12",
               "https://www.mlb.com/stats/2025?page=13", "https://www.mlb.com/stats/2025?page=14", "https://www.mlb.com/stats/2025?page=15", "https://www.mlb.com/stats/2025?page=16",
               "https://www.mlb.com/stats/2025?page=17", "https://www.mlb.com/stats/2025?page=18", "https://www.mlb.com/stats/2025?page=19", "https://www.mlb.com/stats/2025?page=20",
               "https://www.mlb.com/stats/2025?page=21", "https://www.mlb.com/stats/2025?page=22", "https://www.mlb.com/stats/2025?page=23", "https://www.mlb.com/stats/2025?page=24",
               "https://www.mlb.com/stats/2025?page=25", "https://www.mlb.com/stats/2025?page=26", "https://www.mlb.com/stats/2025?page=27", "https://www.mlb.com/stats/2025?page=28",
               "https://www.mlb.com/stats/2025?page=29", "https://www.mlb.com/stats/2025?page=30", "https://www.mlb.com/stats/2025?page=31"]

from selenium.webdriver.common.by import By

mlb_average = [] #initializes empty list called mlb_average
for url in mlb_urls: #loops through each url in list
    driver.get(url) #uses selenium to navigate to page
    driver.implicitly_wait(7) #Tells Selenium to wait up to 7 seconds for elements to appear on the page before throwing an error.
    
    rows = driver.find_elements(By.XPATH, '//*[@id="stats-app-root"]/section/section/div[4]/div[1]/div/table/tbody/tr')
    #finds all table rows insidestats table. finds elemnets and returns a list for each player
    for row in rows: #loops through each row in the list of player rows 
        try:
            avg = row.find_element(By.XPATH, './td[14]').text #Looks inside the current row for the 14th element (the batting average) and extracts the number as a string
            mlb_average.append(float(avg)) #appends to the list mlb_average as a float instead of string
        except:
            continue

In [22]:
mlb_average[0] #check it worked

0.247

In [23]:
#ops

 
chrome_options = Options()
driver = webdriver.Chrome() # Initialize
chrome_options.add_argument("--window-size=1900,1000") #pick a size and stick with it, it can change how code intreacts with the website.
driver = webdriver.Chrome(options = chrome_options) #establish driver
#__________________________________________________________________________________________________#
mlb_urls = ["https://www.mlb.com/stats/2025", "https://www.mlb.com/stats/2025?page=2", "https://www.mlb.com/stats/2025?page=3", 
               "https://www.mlb.com/stats/2025?page=4", "https://www.mlb.com/stats/2025?page=5", "https://www.mlb.com/stats/2025?page=6", 
               "https://www.mlb.com/stats/2025?page=7", "https://www.mlb.com/stats/2025?page=8", "https://www.mlb.com/stats/2025?page=9",
               "https://www.mlb.com/stats/2025?page=10", "https://www.mlb.com/stats/2025?page=11", "https://www.mlb.com/stats/2025?page=12",
               "https://www.mlb.com/stats/2025?page=13", "https://www.mlb.com/stats/2025?page=14", "https://www.mlb.com/stats/2025?page=15", "https://www.mlb.com/stats/2025?page=16",
               "https://www.mlb.com/stats/2025?page=17", "https://www.mlb.com/stats/2025?page=18", "https://www.mlb.com/stats/2025?page=19", "https://www.mlb.com/stats/2025?page=20",
               "https://www.mlb.com/stats/2025?page=21", "https://www.mlb.com/stats/2025?page=22", "https://www.mlb.com/stats/2025?page=23", "https://www.mlb.com/stats/2025?page=24",
               "https://www.mlb.com/stats/2025?page=25", "https://www.mlb.com/stats/2025?page=26", "https://www.mlb.com/stats/2025?page=27", "https://www.mlb.com/stats/2025?page=28",
               "https://www.mlb.com/stats/2025?page=29", "https://www.mlb.com/stats/2025?page=30", "https://www.mlb.com/stats/2025?page=31"]

from selenium.webdriver.common.by import By

mlb_ops = [] #initializes list
for url in mlb_urls: #loops through all urls in list
    driver.get(url) #navigates to page
    driver.implicitly_wait(7) #waits 7 seconds for elemnets to appear before error message
    
    rows = driver.find_elements(By.XPATH, '//*[@id="stats-app-root"]/section/section/div[4]/div[1]/div/table/tbody/tr')
    # finds all table rows inside stats table. finds elements and returns a list for each player
    for row in rows:
        try:
            ops = row.find_element(By.XPATH, './td[17]').text #Looks inside the current row for the 17 element (ops) and extracts the number as a string
            mlb_ops.append(float(ops)) #appends the number as a float to list mlb_ops
        except:
            continue

In [32]:
mlb_ops[0] #checked if it worked

0.948

In [33]:
df1 = pd.DataFrame({ #creates new pandas dataframe
    'Player': players, #creates column called player and uses my list players
    'OPS': mlb_ops, #creates column called OPS and uses my mlb_ops list
    'HR': homers, #creates column called HR and uses my homers list
    'BattingAvg': mlb_average, #creates column called BattingAVG using my mlb_average list
    'RBI': mlb_rbis #creates column called RBI using my mlb_rbi list
})
#I did this way because all my lists are from the same sites which means they are all same lenghth (i double checked before)
#The index allignment is by list order (first element of each list corresponds iwth the same player)
#df1 is a structured table of player stats that can be merged with other dataframes

In [69]:
df1.head()

,Player,OPS,HR,BattingAvg,RBI
0,Cal Raleigh,0.948,60,0.247,125
1,Kyle Schwarber,0.928,56,0.240,132
2,Shohei Ohtani,1.014,55,0.282,102
3,Aaron Judge,1.145,53,0.331,114
4,Eugenio Suárez,0.824,49,0.228,118


In [37]:
#site2

mlb_url2 = ['https://databases.usatoday.com/major-league-baseball-salaries-2025/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/2/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/3/',
            'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/4/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/5/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/6/',
            'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/7/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/8/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/9/',
            'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/10/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/11/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/12/', 
            'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/13/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/14/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/15/',
            'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/16/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/17/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/18/',
            'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/19/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/20/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/21/',
            'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/22/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/23/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/24/', 
            'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/25/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/26/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/27/',
            'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/28/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/29/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/30/',
            'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/31/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/32/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/33/',
            'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/34/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/35/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/36/',
            'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/37/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/38/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/39/',
            'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/40/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/41/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/42/',
            'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/43/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/44/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/45/', 
            'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/46/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/47/', 'https://databases.usatoday.com/major-league-baseball-salaries-2025/page/48/'
           ]

#list of the second website i am scraping (all the pages to it, totl of 48)


In [50]:
#player 2

players2 = [] #empty list called players2. I added this so I can merge this dataframe to first one by player

chrome_options = Options()
driver = webdriver.Chrome() # Initialize
chrome_options.add_argument("--window-size=1900,1000") #pick a size and stick with it, it can change how code intreacts with the website.
driver = webdriver.Chrome(options = chrome_options) #establish driver


for url in mlb_url2: #loops through list mlb_url2
    driver.get(url) #opens url in browser
    driver.implicitly_wait(7) #waits 7 seconds for elemnets to load before error message
    
    rows = driver.find_elements(By.XPATH, '//*[@id="csp-data"]/div[3]/div/div/div[1]/table/tbody/tr')
    #finds all elements in table. returns list of rows. The XPATH points to the <tbody> of the table where each row represents a player.
    for row in rows: #loops through each row in table 
        try:
            mlb_players = row.find_element(By.XPATH, './td[1]').text #finds first cell in current row (player name) and extracts as string
            players2.append(mlb_players) #appends name to players name
        except:
            continue


In [52]:
players2[0]

'Soto, Juan'

In [55]:
#corrected players list
#on the website the names are written as Last, First. I had to get rid of the comma and make the name First Last so it matched with my first df

corrected_players2 = [] #new empty list for corrected format

for name in players2: #loops through original players2 list
    if "," in name: #checks if string has comma and if it does the string needs to be fixed
        parts = name.split(",") #splits string at the comma to 2 parts
        full_name = parts[1].strip() + " " + parts[0].strip() #part[1] is first name and part[0] is the last name. strip removes any extra spaces so the format is in First Last
    else:
        full_name = name.strip() #if name does have comma order is assumed to be in First Last and removes any extra spaces. This is here just for percausion but every name has a comma 
    corrected_players2.append(full_name) # appends full name in correct format to the list corrected_players2


In [57]:
len(corrected_players2) #check it worked 

954

In [58]:
players2 = corrected_players #made players 2 equal the correct list so I can use the list players2

In [59]:

#salary
salary = [] #empty list salary

chrome_options = Options()
driver = webdriver.Chrome() # Initialize
chrome_options.add_argument("--window-size=1900,1000") #pick a size and stick with it, it can change how code intreacts with the website.
driver = webdriver.Chrome(options = chrome_options) #establish driver


for url in mlb_url2: #loops through list specified
    driver.get(url) #opens url in browser using selenium 
    driver.implicitly_wait(7) #waits for data to load before error message (7 seconds)
    
    rows = driver.find_elements(By.XPATH, '//*[@id="csp-data"]/div[3]/div/div/div[1]/table/tbody/tr')
    #finds all elements in table. returns list of rows. The XPATH points to the <tbody> of the table where each row represents a player.
    for row in rows: #loops through each row in table
        try:
            player_sal = row.find_element(By.XPATH, './td[4]').text #finds 4 cell in current row (salary) and extracts as string
            salary.append(player_sal) #adds vaue to empty list
        except:
            continue

In [61]:
salary[0] #check if it worked

'$61,875,000'

In [62]:
#cleaning salary list so it is integer 
#the salary extracted with $ in front of number. I want this to be an integer
salaries_int = [int(s.replace("$","").replace(",","")) for s in salary] #loops through list and replaces the , and $ in each element in the list and stores it to a new list as an integer type


In [64]:
len(salaries_int) #check it worked

954

In [65]:
salary=salaries_int #makes salary list equal the corrected format

In [66]:
# team

team = [] #initializes new list

chrome_options = Options()
driver = webdriver.Chrome() # Initialize
chrome_options.add_argument("--window-size=1900,1000") #pick a size and stick with it, it can change how code intreacts with the website.
driver = webdriver.Chrome(options = chrome_options) #establish driver


for url in mlb_url2: #loops through each website
    driver.get(url) #opens url in browser using selenium 
    driver.implicitly_wait(7)
    
    rows = driver.find_elements(By.XPATH, '//*[@id="csp-data"]/div[3]/div/div/div[1]/table/tbody/tr')
     #finds all elements in table. returns list of rows. The XPATH points to the <tbody> of the table where each row represents a player.
    for row in rows: #loops through each row in table
        try:
            mlb_team = row.find_element(By.XPATH, './td[2]').text #finds 2 cell in current row (team) and extracts as string
            team.append(mlb_team) #appends the string to the list team
        except:
            continue

In [68]:
len(team) #check it worked 

954

In [70]:
#position 

position = [] #initalizes position empty list

chrome_options = Options()
driver = webdriver.Chrome() # Initialize
chrome_options.add_argument("--window-size=1900,1000") #pick a size and stick with it, it can change how code intreacts with the website.
driver = webdriver.Chrome(options = chrome_options) #establish driver


for url in mlb_url2:
    driver.get(url)
    driver.implicitly_wait(7)
    
    rows = driver.find_elements(By.XPATH, '//*[@id="csp-data"]/div[3]/div/div/div[1]/table/tbody/tr')
     #finds all elements in table. returns list of rows. The XPATH points to the <tbody> of the table where each row represents a player.
    for row in rows:
        try:
            mlb_pos = row.find_element(By.XPATH, './td[3]').text #finds 3 cell in current row (position) and extracts as string
            position.append(mlb_pos) #appends to list mlb position
        except:
            continue

In [72]:
len(position) #check it worked

954

In [76]:
#cleaning position list 
#3 position on the website were just values of 1 2 and 3 corresponding to 1 base 2 base and 3 base. I thought this was confusing and wanted to make it more clear by replacing values
clean_pos = position
for i in range(len(clean_pos)): #loops through entire position list
    if clean_pos[i] == '1':
        clean_pos[i] = "1B" #if value = 1 change to 1B
    elif clean_pos[i] == '2':
        clean_pos[i] = "2B" #if value = 2 change to 2B
    elif clean_pos[i] == '3':
        clean_pos[i] = "3B" #if value =3 change to 3B

In [80]:
len(clean_pos) #check it worked

954

In [81]:
# creating df2

df2 = pd.DataFrame({
    "Player": players2,
    "Position": clean_pos,
    "Team": team,
    "Salary": salary
})
#just like before all information was extracted from the same website. I put all my lists into a dataframe saving it as df2. 
#I was able to do this because all my first elements match with the first player and all my lists were same lenghth

In [83]:
df2.head() #check dataframe worked 

,Player,Position,Team,Salary
0,Juan Soto,OF,Mets,61875000
1,Zack Wheeler,RHP,Phillies,42000000
2,Aaron Judge,OF,Yankees,40000000
3,Jacob deGrom,RHP,Rangers,40000000
4,Anthony Rendon,3B,Angels,38571429


In [84]:
# merging 2 dfs
# Merge on Player column
merged_df = pd.merge(df1, df2, on="Player", how="inner")  # only keeps players in both dataframes and all data corresponding to the player is in one line.
#drops players that are only in 1 dataframe but not both

In [85]:
merged_df.head() #check if it worked 

,Player,OPS,HR,BattingAvg,RBI,Position,Team,Salary
0,Cal Raleigh,0.948,60,0.247,125,C,Mariners,2666667
1,Kyle Schwarber,0.928,56,0.240,132,DH,Phillies,20000000
2,Shohei Ohtani,1.014,55,0.282,102,DH/RHP,Dodgers,28213523
3,Aaron Judge,1.145,53,0.331,114,OF,Yankees,40000000
4,Junior Caminero,0.846,45,0.264,110,3B,Rays,764100


In [86]:
len(merged_df) #check lenghth of final dataframe

431

In [ ]:
#the regression I want to run using this data is what effects players salary. I will see how ops,hr,battingavg, and position effect your salary in mlb
#all my stats are from the completed 2025 regular season 
#this regression will provide economic logic that teams ue when they decide on contracts they want to offer
#it will also allow me to incoorportate interactive terms (HR × Position) to see if some positions benefit more from power hitting.
#I used two websites to create this dataset 
#https://www.mlb.com/stats/2025 and https://databases.usatoday.com/major-league-baseball-salaries-2025/
#all my variables were webscraped 

In [87]:
# cvs file 
merged_df.to_csv("mlb_data.csv", index=False)


In [88]:
import os
print(os.getcwd())

C:\Users\lamas\OneDrive\Documents\eco590\assignment-1-Jesslamastra


In [89]:
df_check = pd.read_csv("mlb_data.csv")
print(df_check.head())

            Player    OPS  HR  BattingAvg  RBI Position      Team    Salary
0      Cal Raleigh  0.948  60       0.247  125        C  Mariners   2666667
1   Kyle Schwarber  0.928  56       0.240  132       DH  Phillies  20000000
2    Shohei Ohtani  1.014  55       0.282  102   DH/RHP   Dodgers  28213523
3      Aaron Judge  1.145  53       0.331  114       OF   Yankees  40000000
4  Junior Caminero  0.846  45       0.264  110       3B      Rays    764100


Prof K
We talked about this, but simplify the code so you run the for loop once and then collect everythign you need within that

explain why you use for and try in the code

There is a more efficient way to do this in pandas: 
#cleaning position list 
#3 position on the website were just values of 1 2 and 3 corresponding to 1 base 2 base and 3 base. I thought this was confusing and wanted to make it more clear by replacing values
clean_pos = position
for i in range(len(clean_pos)): #loops through entire position list
    if clean_pos[i] == '1':
        clean_pos[i] = "1B" #if value = 1 change to 1B
    elif clean_pos[i] == '2':
        clean_pos[i] = "2B" #if value = 2 change to 2B
    elif clean_pos[i] == '3':
        clean_pos[i] = "3B" #if value =3 change to 3B
        

you dropped a lot of people in the merges (from 954 to 431) - find out why. this could be a merge issue and you need to figure out why some people aren't merging in correctly.

you mostly wrote this yourself - this give you more points for that.

explain more why you code somet things, not just a quick summary. 


Grade: 90